src/main/java/org/alfresco/repo/security/permissions/impl/SimplePermissionReference.java

```diff
@@ -40,59 +39,32 @@
 {   
     private static final long serialVersionUID = 637302438293417818L;
 
-    private static ReadWriteLock lock = new ReentrantReadWriteLock();
-
-    private static HashMap<QName, HashMap<String, SimplePermissionReference>> instances = new HashMap<QName, HashMap<String, SimplePermissionReference>>();
+    private static HashMap<Pair<QName, String>, SimplePermissionReference> instances = new HashMap<>();

```

Comment:
Just an observation. Wouldn't it be better to replace the HashMap with a ConcurrentHashMap? That could help in avoiding locks and improve scalability.

Reply:
As far as I understood the ConcurrentHashMap provides a copy for all readers while the write operation happens. This means that the change like this https://github.com/Alfresco/alfresco-repository/pull/1189/commits/92c29c01615f7828cebc089904d722761bc8caf2 may result in duplicate objects if parallel threads got null and updated the entry one after another. On the other hand if the write operation is fully synchronised, then there is no reason for ConcurrentHashMap in this case.

Reply:
Even though it doesn't lock on Read operations the Map could be updated atomically using computeIfAbsent function to avoid any duplicates.